File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.java
class TrieNode {
    TrieNode[] children = new TrieNode[26];
    boolean isEnd = false;
}
class Trie {
    TrieNode root;
    public Trie() {
        this.root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode curr = this.root;
        for (Character c : word.toCharArray()) {
            if (curr.children[c - 97] == null) {
                curr.children[c - 97] = new TrieNode();
            }
            curr = curr.children[c - 97];
        }
        curr.isEnd = true;
    }
    
    public boolean search(String word) {
        TrieNode curr = this.root;
        for (Character c : word.toCharArray()) {
            if (curr.children[c - 97] == null) {
                return false;
            }
            curr = curr.children[c - 97];
        }
        // curr.isEnd = true;
        if (curr.isEnd == true) {
            return true;
        }
        return false;
    }
    
    public boolean startsWith(String prefix) {
        TrieNode curr = this.root;
        for (Character c : prefix.toCharArray()) {
            if (curr.children[c - 97] == null) {
                return false;
            }
            curr = curr.children[c - 97];
        }
        // curr.isEnd = true;
        // if (curr.isEnd == true) {
        //     return true;
        // }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.py
class TrieNode:
    def __init__(self):
        self.children = [None] * 26
        self.isEnd = False

class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        curr = self.root
        for i in word:
            idx = ord(i) - ord('a')
            if not curr.children[idx]:
                curr.children[idx] = TrieNode()
            curr = curr.children[idx]
        curr.isEnd = True
        return curr

    def search(self, word: str) -> bool:
        curr = self.root
        for i in word:
            idx = ord(i) - ord('a')
            if not curr.children[idx]:
                return False
            curr = curr.children[idx]
        return curr.isEnd == True
    def startsWith(self, prefix: str) -> bool:
        curr = self.root
        for i in prefix:
            idx = ord(i) - ord('a')
            if not curr.children[idx]:
                return False
            curr = curr.children[idx]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)